cpp/csp/adapters/websockets/WebsocketEndpoint.cpp

```diff
@@ -0,0 +1,173 @@
+{ m_on_open=cb; }
+
+void WebsocketEndpointBase::setOnFailCb(void_cb cb) 
+{ m_on_fail=cb; }
+
+void WebsocketEndpointBase::setOnCloseCb(void_cb cb)
+{ m_on_close=cb; }
+
+void WebsocketEndpointBase::onOpen(websocketpp::connection_hdl)
+{ m_on_open(); }
+void WebsocketEndpointBase::onMessage(websocketpp::connection_hdl, message_ptr msg)
+{ m_on_message(msg->get_payload()); }
+void WebsocketEndpointBase::onFail(websocketpp::connection_hdl)
+{ m_on_fail(); }
+void WebsocketEndpointBase::onClose(websocketpp::connection_hdl)
+{ m_on_close(); }
+
+/*
+WebsocketEndpointTLS -> tls impl
+*/
+WebsocketEndpointTLS::WebsocketEndpointTLS(csp::Dictionary properties)
+: WebsocketEndpointBase(properties) 
+{
+    if (m_properties.get<bool>("verbose_log")) {
+        m_client.set_access_channels(websocketpp::log::alevel::all);
+    } else {
+        m_client.clear_access_channels(websocketpp::log::alevel::all);
+    }
+    m_client.init_asio();
+
+    // need to set these with callbacks
+    m_client.set_open_handler(bind(&WebsocketEndpointBase::onOpen, this, ::_1));
+    m_client.set_message_handler(bind(&WebsocketEndpointBase::onMessage, this, ::_1, ::_2));
+    m_client.set_fail_handler(bind(&WebsocketEndpointBase::onFail, this, ::_1));
+    m_client.set_close_handler(bind(&WebsocketEndpointBase::onClose, this, ::_1));
+    m_client.set_tls_init_handler([this](websocketpp::connection_hdl){
+        auto ctx = websocketpp::lib::make_shared<boost::asio::ssl::context>(boost::asio::ssl::context::tlsv12);
+        boost::system::error_code ec;
+        ctx->set_options(
+            boost::asio::ssl::context::default_workarounds |
+            boost::asio::ssl::context::no_sslv2 |
+            boost::asio::ssl::context::single_dh_use, ec
+        );
+        if (ec) {
+            CSP_THROW(csp::RuntimeException, "Init tls failed: "<< ec);
+        }
+        return ctx;
+    });
+}
+
+void WebsocketEndpointTLS::send(std::string s)
+{
+    websocketpp::lib::error_code ec;
+    m_client.send(m_hdl, s, websocketpp::frame::opcode::value::TEXT, ec);
+}
+
+void WebsocketEndpointTLS::run()
+{
+    auto uri = m_properties.get<std::string>("uri");
+    websocketpp::lib::error_code ec;
+    tls_client::connection_ptr con = m_client.get_connection(uri, ec);
+    const csp::Dictionary &headers = *m_properties.get<DictionaryPtr>("headers");
+
+    for( auto it = headers.begin(); it != headers.end(); ++it )
+    {
+        const std::string key = it.key();
+        const std::string value = headers.get<std::string>( key );
+        con.get()->append_header(key, value);
+    }
+    if (ec) {
+        CSP_THROW(RuntimeException, "could not create connection because: " << ec.message());
+    }
+
+    m_client.connect(con);
+    m_hdl = con->get_handle();
+    m_client.run();
+    m_client.reset();
+}
+
+void WebsocketEndpointTLS::close()
+{
+    websocketpp::lib::error_code ec;
+    m_client.close(m_hdl, websocketpp::close::status::going_away, "Good bye", ec);
```

Comment:
Is the "Good bye" standard or put in for testing?

Reply:
I just chose something, I guess it could be anything

Reply:
the symphony adapter does something similar and I kind of dislike it. Is it possible to configure this? considering its on websocket close, it doesnt really seem like you need to send anything, maybe an empty string is sufficient?

Reply:
I adjusted this to be an empty string. We could always add it as a property (on_close_message or something similar) if there is a use case.